File: DropDownList/MultiSelectDropDown/MultiSelectDropDownCSharp/CustomEditorElement.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Telerik.WinControls.UI;
using Telerik.WinControls.Layouts;
using Telerik.WinControls;
using System.Drawing;
using System.Windows.Forms;
using System.Data;

namespace MultiSelectDropDown
{
    public class CustomEditorElement : RadDropDownListEditorElement
    {
        private LightVisualElement customText;
        private RadButtonElement closeButton;
        private bool textChanged;

        public CustomEditorElement()
        {
            closeButton = new RadButtonElement("Close");
            closeButton.SetValue(DockLayoutPanel.DockProperty, Dock.Bottom);
            closeButton.Click += new EventHandler(closeButton_Click);
            this.Popup.SizingGripDockLayout.Children.Insert(1, closeButton);
            this.SelectionMode = System.Windows.Forms.SelectionMode.MultiSimple;

            this.PopupClosing += new RadPopupClosingEventHandler(CustomEditorElement_PopupClosing);
            this.CreatingVisualItem += new CreatingVisualListItemEventHandler(CustomEditorElement_CreatingVisualItem);
            this.ListElement.ItemDataBinding += this.CustomEditorElement_ItemDataBinding;
        }

        private void deselectAll_Click(object sender, EventArgs e)
        {
            this.SetItemsCheckSelect(false);
        }

        private void selectAll_Click(object sender, EventArgs e)
        {
            this.SetItemsCheckSelect(true);
        }

        private void SetItemsCheckSelect(bool value)
        {
            foreach (CustomListDataItem item in this.Items)
            {
                item.Selected = value;
                item.Checked = value;
            }

            this.SynchronizeText();
        }

        protected override void SyncVisualProperties(RadListDataItem listItem)
        {
        }

        void closeButton_Click(object sender, EventArgs e)
        {
            ClosePopup();
            GridDataCellElement cell = this.Parent as GridDataCellElement;
            if (cell != null)
            {
                cell.GridViewElement.EndEdit();
            }
        }

        private void CustomEditorElement_ItemDataBinding(object sender, ListItemDataBindingEventArgs args)
        {
            args.NewItem = new CustomListDataItem();
        }

        void CustomEditorElement_CreatingVisualItem(object sender, CreatingVisualListItemEventArgs args)
        {
            args.VisualItem = new CustomListVisualItem();
        }

        void CustomEditorElement_PopupClosing(object sender, RadPopupClosingEventArgs args)
        {
            CustomEditorElement editor = (CustomEditorElement)sender;
            if (args.CloseReason == RadPopupCloseReason.Mouse)
            {
                if (editor.PopupForm.Bounds.Contains(Control.MousePosition))
                {
                    args.Cancel = true;
                }
            }
        }

        protected override void CreateChildElements()
        {
            base.CreateChildElements();

            customText = new LightVisualElement();
            customText.DrawBorder = false;
            customText.DrawFill = true;
            customText.GradientStyle = GradientStyles.Solid;
            customText.BackColor = Color.White;
            customText.TextAlignment = ContentAlignment.MiddleLeft;
            this.EditableElement.Children.Add(customText);
            this.TextBox.Visibility = ElementVisibility.Collapsed;
            this.MinSize = new Size(0, 21);
        }

        public override void ShowPopup()
        {
            bool[] selected = new bool[this.Items.Count];
            for (int i = 0; i < selected.Length; i++)
            {
                selected[i] = this.Items[i].Selected;
            }

            base.ShowPopup();
            for (int i = 0; i < selected.Length; i++)
            {
                this.Items[i].Selected = selected[i];
            }
        }

        protected override void OnTextChanged(EventArgs e)
        {
            SynchronizeText();
        }

        internal void SynchronizeText()
        {
            if (textChanged)
            {
                return;
            }

            textChanged = true;
            StringBuilder text = new StringBuilder();
            foreach (CustomListDataItem item in this.ListElement.Items)
            {
                if (item.Checked) 
                {
                    text.AppendFormat("{0}; ", item.Text);
                }
            }

            customText.Text = text.ToString();
            textChanged = false;        
        }
    }
}

File: DropDownList/MultiSelectDropDown/MultiSelectDropDownVB/CustomEditorElement.vb
Imports Telerik.WinControls.UI
Imports Telerik.WinControls.Layouts
Imports Telerik.WinControls
Imports System.Text

Public Class CustomEditorElement
    Inherits RadDropDownListEditorElement
    Private customText As LightVisualElement
    Private closeButton As RadButtonElement
    Private textChanged As Boolean

    Public Sub New()
        closeButton = New RadButtonElement("Close")
        closeButton.SetValue(DockLayoutPanel.DockProperty, Dock.Bottom)
        AddHandler closeButton.Click, AddressOf closeButton_Click
        Me.Popup.SizingGripDockLayout.Children.Insert(1, closeButton)
        Me.SelectionMode = System.Windows.Forms.SelectionMode.MultiSimple

        AddHandler Me.PopupClosing, AddressOf CustomEditorElement_PopupClosing
        AddHandler Me.CreatingVisualItem, AddressOf CustomEditorElement_CreatingVisualItem
        AddHandler Me.ListElement.ItemDataBinding, AddressOf Me.CustomEditorElement_ItemDataBinding
    End Sub

    Private Sub deselectAll_Click(sender As Object, e As EventArgs)
        Me.SetItemsCheckSelect(False)
    End Sub

    Private Sub selectAll_Click(sender As Object, e As EventArgs)
        Me.SetItemsCheckSelect(True)
    End Sub

    Private Sub SetItemsCheckSelect(value As Boolean)
        For Each item As CustomListDataItem In Me.Items
            item.Selected = value
            item.Checked = value
        Next

        Me.SynchronizeText()
    End Sub

    Protected Overrides Sub SyncVisualProperties(listItem As RadListDataItem)
    End Sub

    Private Sub closeButton_Click(sender As Object, e As EventArgs)
        ClosePopup()
        Dim cell As GridDataCellElement = TryCast(Me.Parent, GridDataCellElement)
        If cell IsNot Nothing Then
            cell.GridViewElement.EndEdit()
        End If
    End Sub

    Private Sub CustomEditorElement_ItemDataBinding(sender As Object, args As ListItemDataBindingEventArgs)
        args.NewItem = New CustomListDataItem()
    End Sub

    Private Sub CustomEditorElement_CreatingVisualItem(sender As Object, args As CreatingVisualListItemEventArgs)
        args.VisualItem = New CustomListVisualItem()
    End Sub

    Private Sub CustomEditorElement_PopupClosing(sender As Object, args As RadPopupClosingEventArgs)
        Dim editor As CustomEditorElement = DirectCast(sender, CustomEditorElement)
        If args.CloseReason = RadPopupCloseReason.Mouse Then
            If editor.PopupForm.Bounds.Contains(Control.MousePosition) Then
                args.Cancel = True
            End If
        End If
    End Sub

    Protected Overrides Sub CreateChildElements()
        MyBase.CreateChildElements()

        customText = New LightVisualElement()
        customText.DrawBorder = False
        customText.DrawFill = True
        customText.GradientStyle = GradientStyles.Solid
        customText.BackColor = Color.White
        customText.TextAlignment = ContentAlignment.MiddleLeft
        Me.EditableElement.Children.Add(customText)
        Me.TextBox.Visibility = ElementVisibility.Collapsed
        Me.MinSize = New Size(0, 21)
    End Sub

    Public Overrides Sub ShowPopup()
        Dim selected As Boolean() = New Boolean(Me.Items.Count - 1) {}
        For i As Integer = 0 To selected.Length - 1
            selected(i) = Me.Items(i).Selected
        Next

        MyBase.ShowPopup()
        For i As Integer = 0 To selected.Length - 1
            Me.Items(i).Selected = selected(i)
        Next
    End Sub

    Protected Overrides Sub OnTextChanged(e As EventArgs)
        SynchronizeText()
    End Sub

    Friend Sub SynchronizeText()
        If textChanged Then
            Return
        End If

        textChanged = True
        Dim text As New StringBuilder()
        For Each item As CustomListDataItem In Me.ListElement.Items
            If item.Checked Then
                text.AppendFormat("{0}; ", item.Text)
            End If
        Next

        customText.Text = text.ToString()
        textChanged = False
    End Sub
End Class